src/Extensions/Hosting.Services.Web/Middlewares/SaltProviders/RotatingSaltProvider.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Security.Cryptography;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Omex.Extensions.Hosting.Services.Web.Middlewares
+{
+	internal class RotatingSaltProvider : ISaltProvider
+	{
+		private const int SaltLength = 48; // currently recomended salt size 32
+		private const int HoursToKeepSalt = 40; // should not be more then 48 hours
+		private readonly RandomNumberGenerator m_random;
+		private readonly IMemoryOwner<byte> m_currentSaltMemory;
+		private readonly ISystemClock m_systemClock;
+		private DateTimeOffset m_saltGenerationTime;
+
+		public RotatingSaltProvider(ISystemClock systemClock)
+		{
+			m_random = new RNGCryptoServiceProvider();
+			m_currentSaltMemory = MemoryPool<byte>.Shared.Rent(SaltLength);
+			m_saltGenerationTime = DateTime.MinValue;
```

Comment:
The actual salt provider should be synchronized across all nodes/regions, so that user hashes are the same for the same user in a specific time frame when they hit various VMs. Do I understand correctly that this will only work on a single VM?

Reply:
Yes, with current implementation created salt would be different on each machine (but then passed as part of the request in the case if service would need to make nested calls) since it's the easiest way of ensuring that information is not traceable back to the user.
In the future, we might add another implementation of salt provider that would call some special 
service or storage for proper salt value the same for all instances, but it's much more dangerous.

Reply:
I think this approach is fine for getting this integrated, then update to a salt service later, if necessary.